### PR TITLE
[OTel] Experimental API for metrics

### PR DESCRIPTION
**Pull request by yashykt**
Provide a public experimental API and bazel compatible build target for OpenTelemetry metrics.

Details -
* New `OpenTelemetryPluginBuilder` class that provides the API specified in https://github.com/grpc/proposal/blob/master/A66-otel-stats.md
* The existing `grpc::internal::OpenTelemetryPluginBuilder` class is moved to `grpc::internal::OpenTelemetryPluginBuilderImpl` for disambiguation.
* Renamed `OTel` in some instances to `OpenTelemetry` for consistency.

### CODE REVIEW

File: include/grpcpp/ext/otel_plugin.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  // generic method type to decide whether to record the method name or to
+  // replace it with "other". Non-generic or pre-registered methods remain
+  // unaffected. If not set, by default, generic method names are replaced with
+  // "other" when recording metrics.
+  OpenTelemetryPluginBuilder& SetGenericMethodAttributeFilter(
+      absl::AnyInvocable<bool(absl::string_view /*generic_method*/) const>
+          generic_method_attribute_filter);
+  // Registers a global plugin that acts on all channels and servers running on
+  // the process.
+  // The most common way to use this API is -
```
Comment (markdroth) [inline include/grpcpp/ext/otel_plugin.h:78]: The rest of this comment should probably go above the class.

File: include/grpcpp/ext/otel_plugin.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+absl::string_view OpenTelemetryServerCallDurationInstrumentName();
+absl::string_view
+OpenTelemetryServerCallSentTotalCompressedMessageSizeInstrumentName();
+absl::string_view
+OpenTelemetryServerCallRcvdTotalCompressedMessageSizeInstrumentName();
+
+class OpenTelemetryPluginBuilder {
+ public:
+  OpenTelemetryPluginBuilder();
+  // If `SetMeterProvider()` is not called, no metrics are collected.
```
Comment (markdroth) [inline include/grpcpp/ext/otel_plugin.h:57]: Comments that should appear in our doxygen doc need three slashes instead of two.

File: src/cpp/ext/csm/csm_observability.cc [hunk 3]
```diff
@@ -48,40 +48,46 @@ namespace experimental {
 // CsmObservabilityBuilder
 //
 
+CsmObservabilityBuilder::CsmObservabilityBuilder()
+    : builder_(
+          std::make_unique<grpc::internal::OpenTelemetryPluginBuilderImpl>()) {}
+
+CsmObservabilityBuilder::~CsmObservabilityBuilder() = default;
```
Comment (markdroth) [inline src/cpp/ext/csm/csm_observability.cc:55]: Why is this needed?

File: src/cpp/ext/otel/BUILD [hunk 4]
```diff
@@ -41,6 +41,7 @@ grpc_cc_library(
         "otel_client_filter.h",
         "otel_plugin.h",
         "otel_server_call_tracer.h",
+        "//:include/grpcpp/ext/otel_plugin.h",
```
Comment (markdroth) [inline src/cpp/ext/otel/BUILD:44]: Is it going to cause a problem to have this header in two different targets?

File: src/cpp/ext/otel/otel_plugin.cc [hunk 5]
```diff
@@ -264,4 +264,76 @@ void OpenTelemetryPluginBuilder::BuildAndRegisterGlobal() {
 }
 
 }  // namespace internal
+
+namespace experimental {
+absl::string_view OpenTelemetryClientAttemptStartedInstrumentName() {
```
Comment (markdroth) [inline src/cpp/ext/otel/otel_plugin.cc:269]: We may have already discussed this, but is there a reason why these need to be functions instead of just constexpr constants?

File: include/grpcpp/ext/otel_plugin.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+#include <memory>
+
+#include "absl/functional/any_invocable.h"
+#include "absl/strings/string_view.h"
+#include "opentelemetry/metrics/meter_provider.h"
+
+namespace grpc {
+
+namespace internal {
+class OpenTelemetryPluginBuilderImpl;
```
Comment (markdroth) [inline include/grpcpp/ext/otel_plugin.h:36]: I assume that once we have the plugin option API in place, there will no longer be any need for having two different layers here (the private API wrapped inside of the public API)?

**Review by markdroth — APPROVED** [review]

File: include/grpcpp/ext/otel_plugin.h (hunk 6 above)
Comment (yashykt) [reply include/grpcpp/ext/otel_plugin.h:36]: Yeah, I think once the experimental `CsmObservability` API goes away, we won't need to do this.

File: include/grpcpp/ext/otel_plugin.h (hunk 2 above)
Comment (yashykt) [reply include/grpcpp/ext/otel_plugin.h]: ah, thanks!

File: include/grpcpp/ext/otel_plugin.h (hunk 1 above)
Comment (yashykt) [reply include/grpcpp/ext/otel_plugin.h]: done!

File: src/cpp/ext/csm/csm_observability.cc (hunk 3 above)
Comment (yashykt) [reply src/cpp/ext/csm/csm_observability.cc:55]: The issue is that `csm_observability.h` only has a forward declaration of `grpc::internal::OpenTelemetryPluginBuilderImpl`. Since we are using this forward declared type inside a unique_ptr, the default destructor of the class can not be used. (That would end up invoking the destructor of the unique_ptr from the header which still doesn't know the type definition resulting in an incomplete type error of the form - 
```
error: invalid application of 'sizeof' to an incomplete type 'grpc::internal::OpenTelemetryPluginBuilderImpl'
```
.) 
So, to avoid this, we define the destructor in the cc file where we have the definition of `grpc::internal::OpenTelemetryPluginBuilderImpl`.

File: src/cpp/ext/otel/BUILD (hunk 4 above)
Comment (yashykt) [reply src/cpp/ext/otel/BUILD:44]: This is an internal target that we can change whenever, so it should be fine

File: src/cpp/ext/otel/otel_plugin.cc (hunk 5 above)
Comment (yashykt) [reply src/cpp/ext/otel/otel_plugin.cc:269]: I had tried that some years ago but had ran into issues. Let me try it again and see if it works now.

Comment (yashykt) [reply src/cpp/ext/otel/otel_plugin.cc:273]: pushed a canary change to see if the tests are happy with it. If they are, I'll use it for all.

File: include/grpcpp/ext/otel_plugin.h (hunk 6 above)
Comment (markdroth) [reply include/grpcpp/ext/otel_plugin.h:36]: It seems like we need the CSM plugin option implemented before we can remove the `CsmObservability` API.  And once we have the CSM plugin option implemented, the `CsmObservability` code can use that plugin option in the public API, so it won't need to reach inside the internal API anymore.  Or am I missing something here?

Comment (yashykt) [reply include/grpcpp/ext/otel_plugin.h:36]: I see what you mean. I wasn't originally planning on changing the implementation of `CsmObservability` to use the plugin option, but I like that idea. I will be able to remove the internal API earlier. Will do that. Thanks!

File: src/cpp/ext/otel/otel_plugin.cc (hunk 5 above)
Comment (yashykt) [reply src/cpp/ext/otel/otel_plugin.cc]: seems to be working. Replaced all

File: include/grpcpp/ext/otel_plugin.h [hunk 7]
```diff
@@ -67,8 +53,27 @@ OpenTelemetryServerCallRcvdTotalCompressedMessageSizeInstrumentName();
 /// grpc.server.call.rcvd_total_compressed_message_size
```
Comment (markdroth) [inline include/grpcpp/ext/otel_plugin.h:53]: Probably no need to list the metrics in the comment now that the constants are in the header file directly.

File: src/cpp/ext/otel/otel_plugin.cc [hunk 8]
```diff
@@ -61,26 +61,24 @@ absl::string_view OpenTelemetryTargetKey() { return "grpc.target"; }
 namespace {
 absl::flat_hash_set<std::string> BaseMetrics() {
```
Comment (markdroth) [inline src/cpp/ext/otel/otel_plugin.cc:62]: Does this need to be `std::string` instead of `absl::string_view`?

**Review by markdroth — COMMENTED** [review]
Just a couple of lingering comments -- feel free to address or ignore.

File: src/cpp/ext/otel/otel_plugin.cc (hunk 8 above)
Comment (yashykt) [reply src/cpp/ext/otel/otel_plugin.cc:62]: I want to intentionally store the set as `std::string` instead of `absl::string_view`. The case I'm worried about is when in the future we allow users to provide the set of metric names that they want enabled and they provide non-constant strings

File: include/grpcpp/ext/otel_plugin.h (hunk 7 above)
Comment (yashykt) [reply include/grpcpp/ext/otel_plugin.h]: done
